sql/src/main/java/io/crate/operation/reference/partitioned/PartitionsVersionExpression.java

```diff
@@ -0,0 +1,104 @@
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.reference.partitioned;
+
+import io.crate.Version;
+import io.crate.metadata.PartitionInfo;
+import io.crate.metadata.RowContextCollectorExpression;
+import io.crate.operation.reference.RowCollectNestedObjectExpression;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+public class PartitionsVersionExpression extends RowCollectNestedObjectExpression<PartitionInfo> {
+
+    public PartitionsVersionExpression() {
+        addChildImplementations();
+    }
+
+    private void addChildImplementations() {
+        childImplementations.put(Version.Property.CREATED.toString(),
+            new PartitionVersionCreatedExpression(Version.Property.CREATED.toString()));
+        childImplementations.put(Version.Property.UPGRADED.toString(),
+            new PartitionVersionCreatedExpression(Version.Property.UPGRADED.toString()));
+    }
+
+    @Override
+    public Map<String, Object> value() {
+        Map<String, Object> map = super.value();
+        return getValuesMap(map);
+    }
+
+    static class PartitionVersionCreatedExpression extends RowCollectNestedObjectExpression<PartitionInfo> {
+
+        PartitionVersionCreatedExpression(String property) {
+            addChildImplementations(property);
+        }
+
+        private void addChildImplementations(String property) {
+            childImplementations.put(Version.CRATEDB_VERSION_KEY,
+                new PartitionDetailedVersionExpression(property, Version.CRATEDB_VERSION_KEY));
+            childImplementations.put(Version.ES_VERSION_KEY,
+                new PartitionDetailedVersionExpression(property, Version.ES_VERSION_KEY));
+        }
+
+        @Override
+        public Map<String, Object> value() {
+            Map<String, Object> map = super.value();
+            return getValuesMap(map);
+        }
+    }
+
+    static class PartitionDetailedVersionExpression extends RowContextCollectorExpression<PartitionInfo, Object> {
+
+        private final String property;
+        private final String dbKey;
+
+        PartitionDetailedVersionExpression(String property, String dbKey) {
+            this.property = property;
+            this.dbKey = dbKey;
+        }
+
+        @Override
+        public Object value() {
+            Version version;
+            if (Version.Property.CREATED.toString().equals(property)) {
+                version = row.versionCreated();
+            } else {
+                version = row.versionUpgraded();
+            }
+
+            return version == null ? null : Version.toStringMap(version).get(dbKey);
+        }
+    }
+
+    @Nullable
+    private static Map<String, Object> getValuesMap(Map<String, Object> map) {
```

Comment:
maybe extract this function somewhere so it can be shared? This is now implemented twice